timesketch/lib/kb/graph/entities.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Graph entities."""
+
+# TODO: Add tests.
+
+
+class BaseEntity(object):
+    """Base class for graph entities."""
+
+    def __init__(self, entity):
+        """Initialize. Attributes will be filled out from the node or edge
+        edge dictionaty.
+
+        Args:
+            entity: Neo4j Node or edge dictionary.
+
+        Attributes:
+            id: Neo4j node or edge ID.
+            type: Neo4j node label or edge type.
+        """
+        self.id = entity[u'id']
+        try:
+            self.type = entity[u'labels'][0]
+        except KeyError:
+            self.type = entity[u'type']
+            self.startNode = entity[u'startNode']
+            self.endNode = entity[u'endNode']
+
+        properties = entity[u'properties']
+        for key in properties:
+            setattr(self, key, properties[key])
+
+    @property
+    def human_readable(self):
+        """Format human readable string for the entity.
+
+        Returns:
+            Human readable string.
+        """
+        return NotImplemented
+
+    def to_dict(self):
+        """Create dictionary from the objects attributes.
+
+        Returns:
+            Dictionary with object attributes.
+        """
+        attributes = self.__dict__
+        attributes[u'human_readable'] = self.human_readable
+        return attributes
+
+
+class GenericEntity(BaseEntity):
+    """Generic entity."""
```

Comment:
What about setting a default NEO4J_TYPE here? Or in the BaseEntity class?

This might catch some errors in the registry below, and the line:
registry[klass.NEO4J_TYPE] = klass

Reply:
Done

Reply:
Not at this moment, I might on Monday when some of this is probably going to be re-written :)